### ISSUE TITLE
Automatic backlog expiry mechanism

### ISSUE THREAD

**Issue by rburchell** [ricochet-im/ricochet#514]
Splitting off from #345; it would be nice to have some kind of time-based backlog expiry.

Quoting that discussion:

> The time-based option is more user-friendly and familiar -- many mobile messaging clients implement it. Usually their time ranges from a few minutes to a week, and the setting is per-conversation and sometimes synced with the contact. Disappearing messages are useful -- think "I don't want to keep this on my screen / system for longer than it takes to talk about", or "I run my ricochet 24/7, and I don't want my conversations from days or weeks ago to still be there if it's compromised".

Suggested implementation would be a setting, similar to this:

    Delete messages after [ 7 days ]
                          [ 30 days ]
                          [ 30 days ]

... such that messages that were *seen* (*not* sent or recieved!) >= the set time would be automatically and silently removed. Durations can well be bikeshedded (e.g. maybe even a 1 hour & 1 day setting would make sense).

We could also think about adding additional scenarios like "when I lock my laptop" or "when I switch conversations", but I think those aren't necessary for an initial implementation. I imagine they'd fit into the above UI as being checkboxes after the combo, or something similar.

I imagine that this would _primarily_ be useful for the long-off day if (when?) we had some form of persistent messaging (#168 for that); but it's still useful for people who keep a long-running instance on for instance a laptop, so I think this is worth an issue.

**Comment by wfr** [comment]
@rburchell I like this idea and implemented a first draft:

![GeneralPreferences screenshot](https://i.imgur.com/ofsSG8q.png)

The setting defaults to the maximum, which is 30 days.
There is a "10 seconds" option for testing purposes.
Currently, the expire() method is called after receiving and after sending a message. We could hook  up a QTimer but it's not necessary, in my humble opinion.

>messages that were seen (not sent or recieved!)

Acknowledged. I'm relying on _m_unreadCount_ for this purpose.

Please let me know what you think.

**Comment by wfr** [comment]
The loop in _ConversationModel::expire_ could be optimized from O(n) to O(log2(n)) complexity because the messages are sorted chronologically.
However _n_ is usually so small that I'd choose shorter, more readable code instead.

**Comment by special** [comment]
@wfr awesome! The code is looking good. I'm idly wondering whether passing `INT_MAX` to expire can overflow and cause chaos..

From a broader UX perspective, I think it should use a timer so that expiration will trigger even if new messages aren't being added. Otherwise, a new message arriving might wipe out huge old conversations all at once -- it'll be jarring. I think I would do by having one timer and updating the interval every time unreadCount changes _or_ a message is added or expired. The interval would be the minimal time before the next message expires. Does that seem reasonable?

I'm probably going to have strong opinions on the exact phrasing and durations. I don't know what those are right now, but I'll think on it and try not to bother you unnecessarily :p

I think relying on unreadCount is ok. Ideally at some point we'll also take activity into account for unreadCount, so leaving the window active won't mark messages as read until there's actual user input. But that's a totally separate problem.

Thanks!

**Comment by wfr** [comment]
@special 
>From a broader UX perspective, I think it should use a timer so that expiration will trigger even if new messages aren't being added. Otherwise, a new message arriving might wipe out huge old conversations all at once -- it'll be jarring. I think I would do by having one timer and updating the interval every time unreadCount changes or a message is added or expired. The interval would be the minimal time before the next message expires. Does that seem reasonable?

Good call. The latest commit implements this. Old messages now disappear at the same pace as they arrived. It's neat to watch :) The timer does not activate when there are unread messages.

>I'm probably going to have strong opinions on the exact phrasing and durations.

Feel free to change it. I don't have strong opinions on that :) Note that I consciously chose plurals to reduce the workload for translators, so we only have "minutes", "hours" and "days", but not "minute", "hour" and "day".

All the best

**Comment by wfr** [comment]
To summarize the state of this issue, there's my current working branch:
https://github.com/ricochet-im/ricochet/compare/master...wfr:issue-514-backlog-expiry
